ste/overwritePrompter.go

```diff
@@ -0,0 +1,93 @@
+// Copyright © Microsoft <wastore@microsoft.com>
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ste
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/Azure/azure-storage-azcopy/common"
+)
+
+// asks questions to confirm overwrite, one routine at a time, i.e. only one active question at a time
+type overwritePrompter struct {
+	lock *sync.Mutex
+
+	// whether we should still ask the user for permission to overwrite the destination
+	// if false, the user has already specified a "for all" answer
+	shouldPromptUser bool
+
+	// if the user made a "for all" selection, save the response here
+	savedResponse bool
+}
+
+func (o *overwritePrompter) shouldOverwrite(objectPath string) (shouldOverwrite bool) {
+	// only one routine can ask the question or check the saved response at a time
+	o.lock.Lock()
```

Comment:
Actually, one thing that does mildly concern me. This is a pointer, shouldn't we be doing some nil checking?

Reply:
I think we shouldn't do any check here, and let it panic, as there's no sane way of handing the situation where the lock is absent, and we must learn of the programming error as soon as possible (instead of hiding it with a if-check).

Reply:
That's a fair analysis.

---
In reply to: [312795852](https://github.com/Azure/azure-storage-azcopy/pull/533#discussion_r312795852) [](ancestors = 312795852)